program/src/processor.rs

```diff
@@ -6146,18 +6146,67 @@ impl Processor {
         let mut mango_group = MangoGroup::load_mut_checked(mango_group_ai, program_id)?;
         check_eq!(admin_ai.key, &mango_group.admin, MangoErrorCode::InvalidAdminKey)?;
         check!(admin_ai.is_signer, MangoErrorCode::SignerNecessary)?;
-        msg!("old referral fee params: ref_surcharge_centibps: {} ref_share_centibps: {} ref_mngo_required: {}", mango_group.ref_surcharge_centibps, mango_group.ref_share_centibps, mango_group.ref_mngo_required);
+        msg!("old referral fee params: ref_surcharge_centibps: {} ref_share_centibps: {} ref_mngo_required: {}", mango_group.ref_surcharge_centibps_tier_1, mango_group.ref_share_centibps_tier_1, mango_group.ref_mngo_required);
 
         // TODO - when this goes out, if there are any events on the EventQueue fee logging will be messed up
 
-        mango_group.ref_surcharge_centibps = ref_surcharge_centibps;
-        mango_group.ref_share_centibps = ref_share_centibps;
+        mango_group.ref_surcharge_centibps_tier_1 = ref_surcharge_centibps;
+        mango_group.ref_share_centibps_tier_1 = ref_share_centibps;
         mango_group.ref_mngo_required = ref_mngo_required;
 
         msg!("new referral fee params: ref_surcharge_centibps: {} ref_share_centibps: {} ref_mngo_required: {}", ref_surcharge_centibps, ref_share_centibps, ref_mngo_required);
         Ok(())
     }
 
+    /// Set the `ref_surcharge_centibps`, `ref_share_centibps` for both tiers and `ref_mngo_required` on `MangoGroup`
+    #[inline(never)]
+    fn change_referral_fee_params2(
+        program_id: &Pubkey,
+        accounts: &[AccountInfo],
+        ref_surcharge_centibps_tier_1: u32,
+        ref_share_centibps_tier_1: u32,
+        ref_surcharge_centibps_tier_2: u32,
+        ref_share_centibps_tier_2: u32,
+        ref_mngo_required: u64,
+    ) -> MangoResult {
+        
+        let tier_2_enabled = ref_surcharge_centibps_tier_2 != 0 && ref_share_centibps_tier_2 != 0;
+        if tier_2_enabled {
+            // tier 1 surcharge must be less than tier 2 surcharge so 10k holders get a discount
+            check!(ref_surcharge_centibps_tier_1 < ref_surcharge_centibps_tier_2, MangoErrorCode::InvalidParam)?;
+            // tier 1 share must be greater than tier 1 surcharge so 10k referees pay more then their referrer
+            check!(ref_share_centibps_tier_1 > ref_surcharge_centibps_tier_1, MangoErrorCode::InvalidParam)?;
+            // tier 2 share must be less than tier 2 surcharge so 100k referees don't pay more than base fee
+            check!(ref_share_centibps_tier_2 < ref_surcharge_centibps_tier_2, MangoErrorCode::InvalidParam)?;
```

Comment:
The lack of symmetry to the above condition is surprising!

Explained to myself like this: share_1 and surcharge_1 are for the 10k MNGO level. But share_2 is for referrers that have more than 100k MNGO while surchare_2 is for accounts that have nothing.

Add a condition that share_2 >= share_1?

Don't we want a total ordering where surcharge_1 <= share_1 <= share_2 <= surcharge_2?

Reply:
Added an extra condition